src/main/java/it/pagopa/pn/deliverypush/service/impl/WebhookServiceImpl.java

```diff
@@ -141,54 +152,101 @@ public Mono<ProgressResponseElementDto> consumeEventStream(String xPagopaPnCxId,
-                        if (StringUtils.hasText(newStatus)
-                            && NotificationStatusInt.valueOf(newStatus) != NotificationStatusInt.IN_VALIDATION
-                            && NotificationStatusInt.valueOf(newStatus) != NotificationStatusInt.REFUSED)
-                            eventEntity.setIun(iun);
-                        eventEntity.setNewStatus(newStatus);
-                        eventEntity.setTimelineEventCategory(timelineEventCategory);
-                        eventEntity.setStreamId(stream.getStreamId());
-                        // il requestId ci va sempre, ed è il base64 dello iun
-                        eventEntity.setNotificationRequestId(Base64Utils.encodeToString(iun.getBytes(StandardCharsets.UTF_8)));
-                        return eventEntityDao.save(eventEntity).then();
+                .collectList()
+                .flatMap(l -> {
+                    if (l.isEmpty()) {
+                        return Mono.empty();    // se non ho stream in ascolto, non c'è motivo di fare le query in dynamo
                     }
                     else {
-                        log.info("skipping saving webhook event for stream={} because timelineeventcategory is not in list timelineeventcategory={}", stream.getStreamId(), timelineEventCategory);
-                        return Mono.empty();
+                        return Mono.fromSupplier(() -> retrieveTimeline(iun, timelineId))
+                                .map(timelineData -> new Object() {
+                                    public final List<StreamEntity> streamList = l;
+                                    public final TimelineElementInternal timelineElementInternal = timelineData.getEvent();
+                                    public final String oldStatus = timelineData.getNotificationStatusUpdate().getOldStatus().getValue();
+                                    public final String newStatus = timelineData.getNotificationStatusUpdate().getNewStatus().getValue();
+                                });
                     }
                 })
-                .collectList().then();
+                .flatMapMany(res -> {
+                    Instant timestamp = res.timelineElementInternal.getTimestamp();
+                    String oldStatus = res.oldStatus;
+                    String newStatus = res.newStatus;
+                    String timelineEventCategory = res.timelineElementInternal.getCategory().getValue();
+                    return Flux.fromIterable(res.streamList)
+                            .flatMap(stream -> processEvent(stream, timestamp, oldStatus, newStatus, timelineEventCategory, timelineId, iun));
+                }).collectList().then();
     }
 
+    private Mono<Void> processEvent(StreamEntity stream, Instant timestamp, String oldStatus, String newStatus, String timelineEventCategory, String timelineId, String iun) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.